ocaml/idl/ocaml_backend/gen_api.ml

```diff
@@ -93,18 +107,32 @@ let gen_record_type ~with_module highapi tys =
         | Some default -> sprintf "(Some (%s))" (Datamodel_values.to_ocaml_string ~v2:true default)
       in
       let make_to_field fld =
-        sprintf {|%s = %s_of_rpc (assocer "%s" x %s)|} (field fld) (OU.alias_of_ty fld.DT.ty)
-          (String.concat "_" fld.DT.full_name) (get_default fld)
+        let rpc_field = rpc_field fld in
+        let get_field ty =
+          let of_rpc_fn ty = sprintf "%s_of_rpc" (OU.alias_of_ty ty) in
+          sprintf "(%s (assocer %s x %s))" (of_rpc_fn ty) rpc_field (get_default fld)
+        in
+        let value =
+          let open DT in
+          match fld.ty with
+          | Option ty ->
+            sprintf "(if List.mem_assoc %s x then Some (%s) else None)"
+              rpc_field (get_field ty)
+          | String | Int | Float | Bool | DateTime | Enum _ | Set _ | Map _ | Ref _ | Record _ ->
+            sprintf "(%s)" (get_field fld.ty)
+        in
+        sprintf "%s = %s" (field fld) value
       in
 
       let type_t = sprintf "type %s_t = { %s }" obj_name (map_fields regular_def) in
       let others = if not with_module then
           []
         else [
-          sprintf "let rpc_of_%s_t x = Rpc.Dict [ %s ]" obj_name (map_fields make_of_field);
+          sprintf "let rpc_of_%s_t x = Rpc.Dict (Xapi_stdext_std.Listext.List.unbox_list [ %s ])" obj_name (map_fields make_of_field);
```

Comment:
This is a rather heavy dependency. `Xapi_stdext_monadic.opt.map` and `Xapi_stdext_std.Listext.List.unbox_list` are so small and easy to implement that it could be worth adding them at line 175 with `assocer` and other local helpers, and avoid having to import two big modules for the api (also because they would then be needed in `xen-api-client` as well for no reason)

Reply:
This is currently not failing only because we are getting the dependency transitively, but with further cleanup it could well break again.